pyasdf/util.py

```diff
@@ -118,3 +121,85 @@ def calculate_padding(content_size, pad_blocks, block_size):
     new_size = int((math.ceil(
         float(new_size) / block_size) + 1) * block_size)
     return max(new_size - content_size, 0)
+
+
+class BinaryStruct(object):
+    """
+    A wrapper around the Python stdlib struct module to define a
+    binary struct more like a dictionary than a tuple.
+    """
+    def __init__(self, descr, endian='>'):
+        """
+        Parameters
+        ----------
+        descr : list of tuple
+            Each entry is a pair ``(name, format)``, where ``format``
+            is one of the format types understood by `struct`.
+
+        endian : str, optional
+            The endianness of the struct.  Must be ``>`` or ``<``.
+        """
+        self._fmt = [endian]
+        self._offsets = {}
+        self._names = []
+        i = 0
+        for name, fmt in descr:
+            self._fmt.append(fmt)
+            self._offsets[name] = (i, ('>' + fmt).encode('ascii'))
+            self._names.append(name)
+            i += struct.calcsize(fmt.encode('ascii'))
+        self._fmt = ''.join(self._fmt).encode('ascii')
+        self._size = struct.calcsize(self._fmt)
+
+    @property
+    def size(self):
+        """
+        Return the size of the struct.
+        """
+        return self._size
+
+    def pack(self, **kwargs):
+        """
+        Pack the given arguments, which are given as kwargs, and
+        return the binary struct.
+        """
+        fields = [0] * len(self._names)
+        for key, val in six.iteritems(kwargs):
+            if key not in self._offsets:
+                raise KeyError("No header field '{0}'".format(key))
+            i = self._names.index(key)
+            fields[i] = val
+        return struct.pack(self._fmt, *fields)
+
+    def unpack(self, buff):
+        """
+        Unpack the given binary buffer into the fields.  The result
+        is a dictionary mapping field names to values.
+        """
+        args = struct.unpack_from(self._fmt, buff[:self._size])
+        return dict(izip(self._names, args))
```

Comment:
Nice little recipe.  I wonder if the explicit `pack` and `unpack` are really necessary--this could just use keyword lookup/assignment built into it to do that automatically.  

Maybe to speed things up don't do struct packing/unpacking on every update/lookup.  But just pack into the struct format in the `__bytes__` method :D